packages/unmock-core/src/service/validator.ts

```diff
@@ -0,0 +1,187 @@
+   * @param serviceSchema The top-level (or recursely-called, hence `any`) schema for the service
+   * @returns An object with string-keys leading to the location of matching state for each variable,
+   *          with value being either the Schema defined for that variable, or undefined.
+   */
+  public static spreadStateFromService(
+    serviceSchema: any,
+    statePath: any,
+  ): { [pathKey: string]: any | undefined } {
+    const matches: { [key: string]: any } = {};
+    for (const key of Object.keys(statePath)) {
+      // Traverse along the different paths in the state
+      const scm = serviceSchema[key];
+      if (["string", "number", "boolean"].includes(typeof statePath[key])) {
+        // The value of the current state key is singular (non-object)
+        // If the service schema contains a matching value (schema), store that one
+        // Otherwise recursively look for matching key in nested schema.
+        matches[key] =
+          scm === undefined || scm.properties !== undefined
+            ? this.matchNested(serviceSchema, key)
+            : scm;
+      } else if (
+        typeof statePath[key] === "object" &&
+        Object.keys(statePath[key]).length > 0
+      ) {
+        // Keep digging through the state and service schema
+        if (typeof scm !== "object" || Object.keys(scm).length === 0) {
+          // Nowhere left to traverse, undefined result
+          matches[key] = undefined;
+        } else {
+          const hasProps = scm !== undefined && scm.properties !== undefined;
+          // Recursively look into `properties` if they exist, maintaining the path
+          const properties = this.spreadStateFromService(
+            hasProps ? scm.properties : scm,
+            statePath[key],
+          );
+          matches[key] = hasProps ? { properties } : properties;
+        }
+      } else {
+        // None of the above - this hints at malformed state (i.e. `{ someKey: undefined }`)
+        throw new Error(
+          `${statePath[key]} (object '${JSON.stringify(
+            statePath,
+          )}' with key '${key}') is not an object, string, number or boolean!`,
+        );
+      }
+    }
+    return matches;
+  }
+
+  /**
+   * Given a media type, return mapping between state, service and response.
+   * @param content
+   * @param state
+   */
+  public static getUpdatedStateFromContent(
+    content: MediaType,
+    state: UnmockServiceState,
+  ) {
+    if (content === undefined || content.schema === undefined) {
+      throw new Error(`No schema defined in ${JSON.stringify(content)}!`);
+    }
+    const schema = content.schema as Schema; // We assume '$ref' are already resolved at this stage
+    // For now, ignore $DSL notation and types
+    const responseModsForState = this.spreadStateFromService(
+      schema.properties,
+      state,
+    ) as Record<string, Schema>; // TODO is this the right type?
+    // TODO: Ignore DSL/convert elements
```

Comment:
Future PR: Embed DSL logic, verify given value in `state` against fetched type in `responseMods`